Automattic/jetpack
# Launchpad: Refactor launchpad task completion for future tasks

Related to Automattic/wp-calypso#77278

## Proposed changes:

* Based on @daledupreez suggestions in paYKcK-30u-p2, this PR attempts to put these ideas into play for real. :)
* Decouple task definitions from task registrations.
* Decouple task list definitions from task list registrations.
* Put task definitions and their callbacks in their own file.
* Add a `wpcom_get_launchpad_task_list` function that fetches a single registered task list in lieu of fetching all task lists at once.
* Add a `has_task_lists` method on the `Launchpad_Task_Lists` singleton that checks to see if we have already registered any task lists.
* Refactor `wpcom_mark_launchpad_task_complete`, allowing us to mark a task as complete by directly interfacing with the task definition and the site options table rather than waiting for the task to be registered.
* Fire a Tracks event when marking a task as complete with the new function.
* Initialize deprecated `add_listener_callback`s for existing tasks.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Apply this patch to your sandbox and sandbox `public-api.wordpress.com`
* Visit the dev console and select WP REST API, `wpcom/v2`
* Make a GET request to `/sites/[siteSlug]/launchpad?checklist_slug=[slug]`, replacing [slug] with the following task list slugs:
- newsletter
- link-in-bio
- link-in-bio-tld
- build
- write
- start-writing
- design-first
- free
- keep-building
- videopress
* Make sure the correct tasks are still returned for each list based on the definitions in `launchpad.php`
* Make sure you can still mark tasks as complete when viewing the Launchpad at `/setup/[flow]/launchpad?siteSlug=[siteSlug]`
